fix(linter): support nested extending

camchenry opened the pull request:
- part of https://github.com/oxc-project/oxc/issues/9307

Allows for using the `extends` keyword to chain multiple extended config files together: for example, a package config that extends a base config that extends a config from `node_modules` and so on.

I don't really love what we're having to do here with using both `override_rules` and `ConfigStoreBuilder.with_rule`. I think there is a refactoring opportunity to merge some functionality together here, but I don't want to refactor and optimize before we have a more comprehensive test suite.

camchenry commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9472?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#9472** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9472?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9472?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#9308** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9308?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9217** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9217?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9156** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9156?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9155** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9155?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9153** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9153?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9152** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9152?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=9472).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

shulaoda commented on crates/oxc_linter/src/config/config_builder.rs:
```diff
@@ -109,23 +109,29 @@ impl ConfigStoreBuilder {
+                            // Use `ConfigStoreBuilder` to load extended config files and then apply rules from those
+                            let extended_config_store =
+                                ConfigStoreBuilder::from_oxlintrc(true, extended_config)?;
+                            for rule in extended_config_store.rules {
+                                builder = builder.with_rule(rule);
+                            }
```
Perhaps we can use `with_rules`? Inserting through `extend` can result in the worst possible memory resizing twice.

Not blocking the PR, we can improve it consistently in the future.

```rust
fn extend<T: IntoIterator<Item = (K, V)>>(&mut self, iter: T) {
    // Keys may be already present or show multiple times in the iterator.
    // Reserve the entire hint lower bound if the map is empty.
    // Otherwise reserve half the hint (rounded up), so the map
    // will only resize twice in the worst case.
    let iter = iter.into_iter();
    let reserve = if self.is_empty() {
        iter.size_hint().0
    } else {
        (iter.size_hint().0 + 1) / 2
    };
    self.reserve(reserve);
    iter.for_each(move |(k, v)| {
        self.insert(k, v);
    });
}
```

camchenry replied:
good call, meant to do this originally, should be fixed now in the latest commit.

Sysix commented on crates/oxc_linter/src/config/config_builder.rs:
```diff
@@ -109,23 +109,29 @@ impl ConfigStoreBuilder {
+                            // TODO(refactor): can we merge this together? seems redundant to use `override_rules` and then
+                            // use `ConfigStoreBuilder`, but we don't have a better way of loading rules from config files other than that.
+                            // Use `override_rules` to apply rule configurations and add/remove rules as needed
+                            extended_config.rules.override_rules(&mut builder.rules, &RULES);
```
we are not using the RulesCache anymore and instead passing all rules.

camchenry replied:
good point, I was having some borrow checking issues here earlier, but turned out to be easier to resolve than I thought.

camchenry commented:
### Merge activity

* **Mar 3, 9:36 PM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

camchenry committed:
fix(linter): support nested extending (#9472)

- part of https://github.com/oxc-project/oxc/issues/9307

Allows for using the `extends` keyword to chain multiple extended config files together: for example, a package config that extends a base config that extends a config from `node_modules` and so on.

I don't really love what we're having to do here with using both `override_rules` and `ConfigStoreBuilder.with_rule`. I think there is a refactoring opportunity to merge some functionality together here, but I don't want to refactor and optimize before we have a more comprehensive test suite.
